daphne/src/hpke.rs

```diff
@@ -184,75 +133,39 @@ impl HpkeReceiverConfig {
-            }
-        }
+        let receiver: Hpke<ImplHpkeCrypto> =
+            check_suite(self.config.kem_id, self.config.kdf_id, self.config.aead_id)?;
+        let sk = HpkePrivateKey::new(self.secret_key.clone());
+        let mut ctx = receiver.setup_receiver(enc, &sk, info, None, None, None)?;
+        let plaintext = ctx.open(aad, ciphertext)?;
+        Ok(plaintext)
     }
 
     /// Generate and return a new HPKE receiver context given a HPKE config ID and HPKE KEM.
     pub fn gen(id: u8, kem_id: HpkeKemId) -> Self {
-        use hpke::Serializable;
-
-        match kem_id {
-            HpkeKemId::X25519HkdfSha256 => {
-                let (sk, pk) = kem::X25519HkdfSha256::gen_keypair(&mut rand::thread_rng());
-                HpkeReceiverConfig {
-                    config: HpkeConfig {
-                        id,
-                        kem_id,
-                        kdf_id: HpkeKdfId::HkdfSha256,
-                        aead_id: HpkeAeadId::Aes128Gcm,
-                        public_key: pk.to_bytes().to_vec(),
-                    },
-                    secret_key: sk.to_bytes().to_vec(),
-                }
-            }
-            HpkeKemId::P256HkdfSha256 => {
-                let (sk, pk) = kem::DhP256HkdfSha256::gen_keypair(&mut rand::thread_rng());
-                HpkeReceiverConfig {
-                    config: HpkeConfig {
-                        id,
-                        kem_id,
-                        kdf_id: HpkeKdfId::HkdfSha256,
-                        aead_id: HpkeAeadId::Aes128Gcm,
-                        public_key: pk.to_bytes().to_vec(),
-                    },
-                    secret_key: sk.to_bytes().to_vec(),
-                }
-            }
-            _ => {
-                // We should never get here.
-                unreachable!("Unsupported KEM ({:?})", kem_id);
-            }
+        let kem = match kem_id {
+            HpkeKemId::P256HkdfSha256 => KemAlgorithm::DhKemP256,
+            HpkeKemId::X25519HkdfSha256 => KemAlgorithm::DhKem25519,
+            _ => unreachable!("Unsupported KEM ({:?})", kem_id),
+        };
+        let kdf = KdfAlgorithm::HkdfSha256;
+        let aead = AeadAlgorithm::Aes128Gcm;
+        let generator = Hpke::<ImplHpkeCrypto>::new(Mode::Base, kem, kdf, aead);
+        if let Ok(keypair) = generator.generate_key_pair() {
+            let pk = keypair.public_key();
+            let sk = keypair.private_key();
+            return HpkeReceiverConfig {
+                config: HpkeConfig {
+                    id,
+                    kem_id,
+                    kdf_id: HpkeKdfId::HkdfSha256,
+                    aead_id: HpkeAeadId::Aes128Gcm,
+                    public_key: Vec::from(pk.as_slice()),
+                },
+                secret_key: Vec::from(sk.as_slice()),
+            };
+        } else {
+            unreachable!("bad key generation for KEM ({:?})", kem_id)
```

Comment:
Is this actually "unreachable"? Why not propagate the error here?

Reply:
take a look